typescript/sdk/src/deploy/middleware/TokenBridgeApp.ts

```diff
@@ -1,7 +1,161 @@
 import { TokenBridgeContracts } from '../../middleware';
-import { ChainName } from '../../types';
+import { ChainMap, ChainName } from '../../types';
+import { objMap } from '../../utils/objects';
+
+import {
+  BridgeAdapterConfig,
+  BridgeAdapterType,
+  CircleBridgeAdapterConfig,
+} from './TokenBridgeRouterDeployer';
+
+const CircleBridgeInterface = ICircleBridge__factory.createInterface();
+const CircleBridgeAdapterInterface =
+  CircleBridgeAdapter__factory.createInterface();
 
+interface CircleBridgeMessage<Chain> {
+  chain: Chain;
+  remoteChain: Chain;
+  txHash: string;
+  message: string;
+  nonce: number;
+  domain: number;
+  nonceHash: string;
+}
 export class TokenBridgeApp<
   Chain extends ChainName = ChainName,
-> extends HyperlaneApp<TokenBridgeContracts, Chain> {}
+> extends HyperlaneApp<TokenBridgeContracts, Chain> {
+  constructor(
+    public readonly contractsMap: ChainMap<Chain, TokenBridgeContracts>,
+    public readonly multiProvider: MultiProvider<Chain>,
+    public readonly bridgeAdapterConfigs: ChainMap<
+      Chain,
+      BridgeAdapterConfig[]
+    >,
+  ) {
+    super(contractsMap, multiProvider);
+  }
+
+  circleBridgeAdapterConfig() {
+    return objMap(
+      this.bridgeAdapterConfigs,
+      (_chain, config) =>
+        config.find(
+          (_) => _.type === BridgeAdapterType.Circle,
+        ) as CircleBridgeAdapterConfig,
+    );
+  }
+
+  async fetchCircleMessageTransactions(chain: Chain): Promise<string[]> {
+    const cc = this.multiProvider.getChainConnection(chain);
+    const params = new URLSearchParams({
+      module: 'logs',
+      action: 'getLogs',
+      address: this.getContracts(chain).circleBridgeAdapter!.address,
+      topic0: CircleBridgeAdapterInterface.getEventTopic(
+        CircleBridgeAdapterInterface.getEvent('BridgedToken'),
+      ),
+    });
+    const req = await fetch(`${cc.getApiUrl()}?${params}`);
+    const response = await req.json();
+
+    return response.result.map((_: any) => _.transactionHash).flat();
+  }
+
+  async parseCircleMessages(
+    chain: Chain,
+    txHash: string,
+  ): Promise<CircleBridgeMessage<Chain>[]> {
+    const connection = this.multiProvider.getChainConnection(chain);
+    const receipt = await connection.provider.getTransactionReceipt(txHash);
+    const matchingLogs = receipt.logs
+      .map((_) => {
+        try {
+          return [CircleBridgeInterface.parseLog(_)];
+        } catch {
+          try {
+            return [CircleBridgeAdapterInterface.parseLog(_)];
+          } catch {
+            return [];
+          }
+        }
+      })
+      .flat();
+
+    if (matchingLogs.length == 0) return [];
+    const message = matchingLogs.find((_) => _!.name === 'MessageSent')!.args
+      .message;
+    const nonce = matchingLogs.find((_) => _!.name === 'BridgedToken')!.args
+      .nonce;
+    const remoteChain =
+      message.chain === Chains.fuji ? Chains.goerli : Chains.fuji;
```

Comment:
this is a little messy
eg we could filter for these in the log topics
happy to accept for MVP

Reply:
Yeah you are right eventually they should be probably parsed from the Dispatch even